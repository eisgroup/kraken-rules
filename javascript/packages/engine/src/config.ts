/*
 *  Copyright 2020 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { debug } from './debugger/Debugger'

/**
 * Configures Kraken evaluation. It can be configured from the console.
 * Open the console and  modify the config options
 * @example
 * global.Kraken.debug = false;
 * global.Kraken.info = false;
 * global.Kraken.debug = false;
 * @see Logger
 * @since 1.1.1
 */
export interface KrakenConfig {
    logger: {
        enabled: boolean
        debug: boolean
        info: boolean
        error: boolean
        warning: boolean
        entriesCount: number
        logs: string[]
        getLogs(): void
        clear(): void
    }
    debugger: debug.api.Debugger
}
export type WithKraken = {
    Kraken: KrakenConfig
}

function init(): void {
    const logger: KrakenConfig['logger'] = {
        enabled: process.env.NODE_ENV !== 'production',
        debug: false,
        info: true,
        error: true,
        warning: true,
        entriesCount: 2000,
        logs: [],
        clear(): void {
            const config = (global as unknown as WithKraken).Kraken
            if (config.logger.logs) {
                config.logger.logs = []
            }
        },
        getLogs(): void {
            const config = (global as unknown as WithKraken).Kraken
            // applicable only for the browser
            if (config.logger.logs && document && window) {
                const blob = new Blob([config.logger.logs.join('\n')], { type: 'text/plain' })
                const url = window.URL.createObjectURL(blob)
                const a = document.createElement('a')
                a.style.display = 'none'
                a.href = url
                a.download = `kraken-logs-${new Date().getTime()}.txt`
                document.body.appendChild(a)
                a.click()
                window.URL.revokeObjectURL(url)
                document.body.removeChild(a)
            }
        },
    }
    if (global) {
        ;(global as unknown as WithKraken).Kraken = {
            logger,
            debugger: new debug.impl.DevToolsDebugger(),
        }
        return
    }
}

function getConfig(): KrakenConfig {
    if (global) {
        return (global as unknown as { Kraken: KrakenConfig })['Kraken']
    }
    throw new Error('Kraken configuration was not initialized')
}

export const krakenConfig = { init, getConfig }
