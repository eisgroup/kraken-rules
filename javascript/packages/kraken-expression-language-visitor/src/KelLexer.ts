import { Common } from './antlr/generated/Common'

export class KelLexer extends Common {
    /** @override */
    popMode(): number {
        return this._modeStack.isEmpty ? Common.DEFAULT_MODE : super.popMode()
    }
}
