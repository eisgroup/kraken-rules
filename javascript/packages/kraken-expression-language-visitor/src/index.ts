export { createParser } from "./parse";
export { KelVisitor } from "./antlr/generated/KelVisitor";
export { SyntaxError } from "./KrakenParser";
export * from "./antlr/generated/Kel";

export { ParserRuleContext } from "antlr4ts";
export { AbstractParseTreeVisitor } from "antlr4ts/tree/AbstractParseTreeVisitor";
export { ErrorNode } from "antlr4ts/tree/ErrorNode";
export { Interval } from "antlr4ts/misc/Interval";
