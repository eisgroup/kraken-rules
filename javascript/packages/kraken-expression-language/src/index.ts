export * from './scope/Scope'
export * from './scope/ScopeType'
export * from './visitor/Node'
export * from './visitor/AstGeneratingVisitor'
export * from './visitor/NodeUtils'

export { KelParser, SyntaxError } from 'kraken-expression-language-visitor'

export * from './visitor/autocomplete/CompletionItem'
export * from './visitor/type-at-location/LocationInfo'

export * from './public/Cache'
export * from './public/KelService'
export * from './public/ScopeDeserializer'

export { Type } from './type/Types'
