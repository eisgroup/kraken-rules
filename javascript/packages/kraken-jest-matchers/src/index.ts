/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { matchers as dateTimeMatchers } from './dateTimeMatchers'
import { matchers as ruleEvalResultMatchers } from './ruleEvalResultMatchers'
import { matchers as entryPointResultMatchers } from './entrypointResultMatchers'

export const matchers = {
    ...dateTimeMatchers,
    ...ruleEvalResultMatchers,
    ...entryPointResultMatchers,
}

export { ExpectedResult } from './entrypointResultMatchers'
export { MatchableResults } from './entrypointResultMatchers'

expect.extend(matchers)
