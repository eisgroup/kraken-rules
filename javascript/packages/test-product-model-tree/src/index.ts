export * from './generated/kraken_model_tree_Policy'
export * from './generated/kraken_model_tree_PolicyExtended'