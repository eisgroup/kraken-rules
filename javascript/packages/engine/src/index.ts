/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import { krakenConfig } from './config'
krakenConfig.init()

// spi
export * from './engine/contexts/info/DataObjectInfoResolver'
export * from './engine/contexts/info/ContextInstanceInfoResolver'

// engine
export * from './engine/executer/SyncEngine'

export * from './bundle-cache/EntryPointBundleCache'
export * from './bundle-cache/delta-cache/RepoClientCache'
export * from './bundle-cache/delta-cache/DeltaBundleCache'
export * from './bundle-cache/dimension-set-cache/DimensionSetBundleCache'
export * from './bundle-cache/dimension-set-cache/DimensionSetCacheLogger'
export * from './bundle-cache/expression-context-manager/ExpressionContextManager'
export * from './bundle-cache/expression-context-manager/ExpressionContextManagerImpl'

export { registry, KelFunction, FunctionParameter } from './engine/runtime/expressions/ExpressionEvaluator'
export { FunctionScope } from './engine/runtime/expressions/functionLibrary/Registry'
export * from './engine/runtime/EvaluationMode'

// Payload result utils
export * from './engine/results/PayloadResultCreator'
export * from './engine/results/PayloadResultTypeChecker'
export { ValueChangedEvent } from './engine/results/ValueChangedEvent'

// Condition result utils
export { conditionEvaluationTypeChecker } from './dto/DefaultConditionEvaluationResult'

export { DataContext } from './engine/contexts/data/DataContext'

// field metadata reducer
export * from './engine/results/field_metadata_reducer/FieldErrorMessage'

export * from './models/EntryPointBundle'
export * from './models/ContextModelTree'

// reducer
export * from './engine/results/RuleOverrideContextExtractor'
export * from './engine/results/Reducer'
export * from './engine/results/Localization'

// field metadata reducer
export * from './engine/results/field_metadata_reducer/FieldMetadataReducer'

// error
export * from './error/KrakenRuntimeError'

// re-export
export {
    ContextInstanceInfo,
    RuleOverride,
    RuleInfo,
    ContextFieldInfo,
    EntryPointResult,
    ConditionEvaluationResult,
    FieldEvaluationResult,
    RuleEvaluationResults,
    PayloadResultType,
    PayloadResult,
    ErrorAwarePayloadResult,
    ErrorMessage,
    ValidationPayloadResult,
    AccessibilityPayloadResult,
    VisibilityPayloadResult,
    DefaultValuePayloadResult,
    AssertionPayloadResult,
    SizePayloadResult,
    SizeRangePayloadResult,
    LengthPayloadResult,
    RegExpPayloadResult,
    UsagePayloadResult,
    ValueListPayloadResult,
    RuleEvent,
    FieldMetadata,
    FieldMetadataResult,
    ConditionEvaluation,
} from 'kraken-engine-api'
