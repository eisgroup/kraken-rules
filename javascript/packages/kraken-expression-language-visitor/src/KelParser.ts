import {
    ANTLRErrorListener,
    CharStreams,
    CommonTokenStream,
    DefaultErrorStrategy,
    InputMismatchException,
    NoViableAltException,
    Parser,
    RecognitionException,
    Recognizer,
    RuleContext,
    Token,
    Vocabulary,
} from 'antlr4ts'

import { ExpressionContext, Kel, TypeContext } from './antlr/generated/Kel'
import { KelLexer } from './KelLexer'

/**
 * Parses KEL expression to tree based structure with antlr4 generated context.
 * This tree can be traversed with visitor {@link KelVisitor} implementation.
 */
export class KelParser {
    private parserEl: KelParserErrorListener
    private lexerEl: KelLexerErrorListener
    private parser: Kel

    constructor(expression: string) {
        const inputStream = CharStreams.fromString(expression)

        const lexer = new KelLexer(inputStream)
        const lexerErrorListener = new KelLexerErrorListener()
        lexer.removeErrorListeners()
        lexer.addErrorListener(lexerErrorListener)
        const tokenStream = new CommonTokenStream(lexer)

        const kelParser = new Kel(tokenStream)
        const parserErrorListener = new KelParserErrorListener()
        kelParser.removeErrorListeners()
        kelParser.addErrorListener(parserErrorListener)

        // error strategy
        kelParser.errorHandler = new ErrorParsingStrategy()

        this.parser = kelParser
        this.parserEl = parserErrorListener
        this.lexerEl = lexerErrorListener
    }
    getSyntaxErrors(): SyntaxError[] {
        return [...this.lexerEl.getSyntaxErrors(), ...this.parserEl.getSyntaxErrors()]
    }
    parseExpression(): ExpressionContext {
        return this.parser.expression()
    }
    parseType(): TypeContext {
        return this.parser.type()
    }
}

export class KelParserErrorListener implements ANTLRErrorListener<Token> {
    private errors: SyntaxError[] = []

    syntaxError(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        _recognizer: Recognizer<Token, any>,
        offendingSymbol: Token | undefined,
        line: number,
        charPositionInLine: number,
        msg: string,
        e: RecognitionException | undefined,
    ): void {
        let os
        if (offendingSymbol) {
            os = {
                line: offendingSymbol.line,
                startIndex: offendingSymbol.startIndex,
                stopIndex: offendingSymbol.stopIndex,
            }
        } else {
            os = {
                line: line,
                startIndex: charPositionInLine,
                stopIndex: charPositionInLine,
            }
        }
        this.errors.push({
            context: e?.context,
            message: msg,
            offendingSymbol: os,
        })
    }

    getSyntaxErrors(): SyntaxError[] {
        return this.errors
    }
}

export class KelLexerErrorListener implements ANTLRErrorListener<number> {
    private errors: SyntaxError[] = []

    syntaxError(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        _recognizer: Recognizer<number, any>,
        _offendingSymbol: number | undefined,
        line: number,
        charPositionInLine: number,
        msg: string,
        e: RecognitionException | undefined,
    ): void {
        const os = {
            line: line,
            startIndex: charPositionInLine,
            stopIndex: charPositionInLine,
        }
        this.errors.push({
            context: e?.context,
            message: msg,
            offendingSymbol: os,
        })
    }

    getSyntaxErrors(): SyntaxError[] {
        return this.errors
    }
}

export interface OffendingSymbol {
    line: number
    startIndex: number
    stopIndex: number
}

export interface SyntaxError {
    context?: RuleContext
    message: string
    offendingSymbol: OffendingSymbol
}

class ErrorParsingStrategy extends DefaultErrorStrategy {
    reportInputMismatch(recognizer: Parser, e: InputMismatchException): void {
        const offendingToken = e.getOffendingToken(recognizer)
        const invalidTokenText = offendingToken?.text
        const expected = e.expectedTokens

        let expectedTokens: string[] = []

        for (const interval of expected?.intervals ?? []) {
            const { a, b } = interval
            if (a === b) {
                const l = this.resolveText(recognizer.vocabulary, a)
                if (l) {
                    expectedTokens = expectedTokens.concat(l)
                }
            } else {
                for (let i = a; i <= b; i++) {
                    // expected token should not be same as offending token
                    if (i !== offendingToken?.type) {
                        continue
                    }
                    const intervalLiterals = this.resolveText(recognizer.vocabulary, i)
                    if (intervalLiterals) {
                        expectedTokens = expectedTokens.concat(intervalLiterals)
                    }
                }
            }
        }
        let message = ''
        if (offendingToken && offendingToken.type === Kel.EOF) {
            message = 'Unexpected end of expression'
        } else if (!expectedTokens.length) {
            message = 'Syntax error occurred, please check the correctness of an expression'
        } else {
            message = `Invalid token '${invalidTokenText}'. Expected ${expectedTokens.join(', ')}`
        }

        recognizer.notifyErrorListeners(message, offendingToken ?? null, e)
    }

    private resolveText(vocabulary: Vocabulary, tokenType: number): string[] {
        const literal = vocabulary.getLiteralName(tokenType)
        if (literal) {
            return [literal]
        }

        const q = (...t: string[]) => t.map(text => `'${text}'`)
        switch (tokenType) {
            case Kel.OP_IN:
            case Kel.IN:
                return q('in')
            case Kel.DOT:
                return q('.')
            case Kel.ELSE:
                return q('else')
            case Kel.EVERY:
                return q('every')
            case Kel.FALSE:
                return q('false')
            case Kel.FOR:
                return q('for')
            case Kel.IF:
                return q('if')
            case Kel.MATCHES:
                return q('matches')
            case Kel.NOT:
                return q('not')
            case Kel.NULL:
                return q('null')
            case Kel.OP_ADD:
                return q('+')
            case Kel.OP_AND:
                return q('and', '&&')
            case Kel.OP_DIV:
                return q('/')
            case Kel.OP_EQUALS:
                return q('=', '==')
            case Kel.OP_EXP:
                return q('**')
            case Kel.OP_INSTANCEOF:
                return q('instanceof')
            case Kel.OP_LESS:
                return q('<')
            case Kel.OP_LESS_EQUALS:
                return q('<=')
            case Kel.OP_MINUS:
                return q('-')
            case Kel.OP_MOD:
                return q('%')
            case Kel.OP_MORE:
                return q('>')
            case Kel.OP_MORE_EQUALS:
                return q('>=')
            case Kel.OP_MULT:
                return q('*')
            case Kel.OP_NEGATION:
                return q('not', '!')
            case Kel.OP_NOT_EQUALS:
                return q('!=')
            case Kel.OR:
            case Kel.OP_OR:
                return q('or', '||')
            case Kel.OP_PIPE:
                return q('|')
            case Kel.OP_QUESTION:
                return q('?')
            case Kel.OP_TYPEOF:
                return q('typeof')
            case Kel.RETURN:
                return q('return')
            case Kel.SATISFIES:
                return q('satisfies')
            case Kel.THIS:
                return q('this')
            case Kel.TRUE:
                return q('true')
        }
        return []
    }

    reportNoViableAlternative(recognizer: Parser, e: NoViableAltException): void {
        const tokens = recognizer.inputStream
        let input
        if (tokens) {
            if (e.startToken.type === Kel.EOF) {
                input = ''
            } else {
                input = tokens.getTextFromRange(e.startToken, e.getOffendingToken())
            }
        } else {
            input = ''
        }

        const message = `Invalid token ${this.escapeWSAndQuote(input)}`
        this.notifyErrorListeners(recognizer, message, e)
    }
}
