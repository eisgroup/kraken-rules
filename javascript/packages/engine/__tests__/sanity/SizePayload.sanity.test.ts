/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { sanityEngine } from './_SanityEngine'
import { sanityMocks } from './_AutoPolicyObject.mocks'
import { FieldMetadataReducer } from '../../src'

describe('Size payload sanity test', () => {
    it('should success with array of 2', () => {
        const policy = sanityMocks.empty()
        policy.policies = ['1', '2']
        const result = sanityEngine.evaluate(policy, 'SizePayload')
        const fieldResults = new FieldMetadataReducer().reduce(result)

        expect(result).k_toMatchResultsStats({ total: 3, critical: 0 })
        expect(result).k_toHaveExpressionsFailures(0)
        expect({ entryPointResults: result, fieldResults }).k_toMatchResultsSnapshots()
    })
    it('should fail with array of 3', () => {
        const policy = sanityMocks.empty()
        policy.policies = ['1', '2', '3']
        const result = sanityEngine.evaluate(policy, 'SizePayload')
        const fieldResults = new FieldMetadataReducer().reduce(result)

        expect(result).k_toMatchResultsStats({ total: 3, critical: 2 })
        expect(result).k_toHaveExpressionsFailures(0)
        expect({ entryPointResults: result, fieldResults }).k_toMatchResultsSnapshots()
    })
    it('should handle undefined collection', () => {
        const policy = sanityMocks.empty()
        policy.policies = undefined
        const result = sanityEngine.evaluate(policy, 'SizePayload')
        const fieldResults = new FieldMetadataReducer().reduce(result)

        expect(result).k_toMatchResultsStats({ total: 3, critical: 2 })
        expect(result).k_toHaveExpressionsFailures(0)
        expect({ entryPointResults: result, fieldResults }).k_toMatchResultsSnapshots()
    })
})
