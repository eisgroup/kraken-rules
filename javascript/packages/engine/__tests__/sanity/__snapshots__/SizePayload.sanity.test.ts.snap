// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Size payload sanity test should fail with array of 3 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "Policy:0:policies": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "0",
          "contextName": "Policy",
          "fieldName": "policies",
          "fieldPath": "policies",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "min": 2,
              "sizeOrientation": "EQUALS",
              "success": false,
              "type": 2,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "SIZE",
              "ruleName": "SizePayloadEquals",
              "targetPath": "policies",
            },
          },
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "min": 2,
              "sizeOrientation": "MIN",
              "success": true,
              "type": 2,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "SIZE",
              "ruleName": "SizePayloadMin",
              "targetPath": "policies",
            },
          },
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "min": 2,
              "sizeOrientation": "MAX",
              "success": false,
              "type": 2,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "SIZE",
              "ruleName": "SizePayloadMax",
              "targetPath": "policies",
            },
          },
        ],
      },
    },
  },
  "fieldResults": Object {
    "Policy:0:policies": Object {
      "fieldType": "STRING",
      "id": "Policy:0:policies",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "policies",
      "ruleResults": Array [
        Object {
          "errorCode": "rule-size-error",
          "errorMessage": "Array length is invalid",
          "isFailed": true,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "min": 2,
            "sizeOrientation": "EQUALS",
            "success": false,
            "type": 2,
            "validationSeverity": "critical",
          },
          "rawTemplateVariables": Array [],
          "ruleName": "SizePayloadEquals",
          "severity": "critical",
          "templateVariables": Array [],
        },
        Object {
          "errorCode": "rule-size-error",
          "errorMessage": "Array length is invalid",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "min": 2,
            "sizeOrientation": "MIN",
            "success": true,
            "type": 2,
            "validationSeverity": "critical",
          },
          "rawTemplateVariables": Array [],
          "ruleName": "SizePayloadMin",
          "severity": "critical",
          "templateVariables": Array [],
        },
        Object {
          "errorCode": "rule-size-error",
          "errorMessage": "Array length is invalid",
          "isFailed": true,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "min": 2,
            "sizeOrientation": "MAX",
            "success": false,
            "type": 2,
            "validationSeverity": "critical",
          },
          "rawTemplateVariables": Array [],
          "ruleName": "SizePayloadMax",
          "severity": "critical",
          "templateVariables": Array [],
        },
      ],
    },
  },
}
`;

exports[`Size payload sanity test should handle undefined collection 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "Policy:0:policies": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "0",
          "contextName": "Policy",
          "fieldName": "policies",
          "fieldPath": "policies",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "min": 2,
              "sizeOrientation": "EQUALS",
              "success": false,
              "type": 2,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "SIZE",
              "ruleName": "SizePayloadEquals",
              "targetPath": "policies",
            },
          },
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "min": 2,
              "sizeOrientation": "MIN",
              "success": false,
              "type": 2,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "SIZE",
              "ruleName": "SizePayloadMin",
              "targetPath": "policies",
            },
          },
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "min": 2,
              "sizeOrientation": "MAX",
              "success": true,
              "type": 2,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "SIZE",
              "ruleName": "SizePayloadMax",
              "targetPath": "policies",
            },
          },
        ],
      },
    },
  },
  "fieldResults": Object {
    "Policy:0:policies": Object {
      "fieldType": "STRING",
      "id": "Policy:0:policies",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "policies",
      "ruleResults": Array [
        Object {
          "errorCode": "rule-size-error",
          "errorMessage": "Array length is invalid",
          "isFailed": true,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "min": 2,
            "sizeOrientation": "EQUALS",
            "success": false,
            "type": 2,
            "validationSeverity": "critical",
          },
          "rawTemplateVariables": Array [],
          "ruleName": "SizePayloadEquals",
          "severity": "critical",
          "templateVariables": Array [],
        },
        Object {
          "errorCode": "rule-size-error",
          "errorMessage": "Array length is invalid",
          "isFailed": true,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "min": 2,
            "sizeOrientation": "MIN",
            "success": false,
            "type": 2,
            "validationSeverity": "critical",
          },
          "rawTemplateVariables": Array [],
          "ruleName": "SizePayloadMin",
          "severity": "critical",
          "templateVariables": Array [],
        },
        Object {
          "errorCode": "rule-size-error",
          "errorMessage": "Array length is invalid",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "min": 2,
            "sizeOrientation": "MAX",
            "success": true,
            "type": 2,
            "validationSeverity": "critical",
          },
          "rawTemplateVariables": Array [],
          "ruleName": "SizePayloadMax",
          "severity": "critical",
          "templateVariables": Array [],
        },
      ],
    },
  },
}
`;

exports[`Size payload sanity test should success with array of 2 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "Policy:0:policies": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "0",
          "contextName": "Policy",
          "fieldName": "policies",
          "fieldPath": "policies",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "min": 2,
              "sizeOrientation": "EQUALS",
              "success": true,
              "type": 2,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "SIZE",
              "ruleName": "SizePayloadEquals",
              "targetPath": "policies",
            },
          },
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "min": 2,
              "sizeOrientation": "MIN",
              "success": true,
              "type": 2,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "SIZE",
              "ruleName": "SizePayloadMin",
              "targetPath": "policies",
            },
          },
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "min": 2,
              "sizeOrientation": "MAX",
              "success": true,
              "type": 2,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "SIZE",
              "ruleName": "SizePayloadMax",
              "targetPath": "policies",
            },
          },
        ],
      },
    },
  },
  "fieldResults": Object {
    "Policy:0:policies": Object {
      "fieldType": "STRING",
      "id": "Policy:0:policies",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "policies",
      "ruleResults": Array [
        Object {
          "errorCode": "rule-size-error",
          "errorMessage": "Array length is invalid",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "min": 2,
            "sizeOrientation": "EQUALS",
            "success": true,
            "type": 2,
            "validationSeverity": "critical",
          },
          "rawTemplateVariables": Array [],
          "ruleName": "SizePayloadEquals",
          "severity": "critical",
          "templateVariables": Array [],
        },
        Object {
          "errorCode": "rule-size-error",
          "errorMessage": "Array length is invalid",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "min": 2,
            "sizeOrientation": "MIN",
            "success": true,
            "type": 2,
            "validationSeverity": "critical",
          },
          "rawTemplateVariables": Array [],
          "ruleName": "SizePayloadMin",
          "severity": "critical",
          "templateVariables": Array [],
        },
        Object {
          "errorCode": "rule-size-error",
          "errorMessage": "Array length is invalid",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "min": 2,
            "sizeOrientation": "MAX",
            "success": true,
            "type": 2,
            "validationSeverity": "critical",
          },
          "rawTemplateVariables": Array [],
          "ruleName": "SizePayloadMax",
          "severity": "critical",
          "templateVariables": Array [],
        },
      ],
    },
  },
}
`;
